### PR TITLE
Feature/494 add cessation cost to offshore facilities

### PR DESCRIPTION
**Pull request by ManveerPandher**
Background for case:
* Within the Datamodels, there's a "Cessation" cost. This is the cost for de-assembling different assets within a project, which also has to be added within the full cost.
* This has been added to Surf, Topside, Substructure and Transport.

What has been done:
* Added CessationCostProfile, within Surf, Transport, SUbstructure and Transport.
* It's similar to CostProfile which each of the Assets has.
* Implemented it within the entire Value-Chain
* Added CessasionCostProfile values to be exported within the Excel Capex fields.

Why this needs to be done:
* Acceptance criteria:
Given that there is a project and a case with asset surf/topside/substructure/transport, when the engineer have calculated cessation costs for an asset then it should be possible to add the cost profiles to the individual assets.

* Given that there is a project and a case with asset surf/topside/substructure/transport, when the engineer add cessation costs to an asset it should be included in the CAPEX/case.

* Given that cessation costs have been added to an asset, when the concept architect want to export data to STEA the cessation cost profile should be included in the export.

Github Case:
* https://github.com/equinor/dcd/issues/494

### CODE REVIEW

File: backend/api/Adapters/SubstructureAdapter.cs [hunk 1]
```diff
@@ -15,7 +15,16 @@ public static Substructure Convert(SubstructureDto substructureDto)
                 DryWeight = substructureDto.DryWeight,
                 Maturity = substructureDto.Maturity
             };
-            substructure.CostProfile = Convert(substructureDto.CostProfile, substructure);
+            if (substructureDto.CostProfile != null)
+            {
+                substructure.CostProfile = Convert(substructureDto.CostProfile, substructure);
+            }
```
Comment (taustad) [inline backend/api/Adapters/SubstructureAdapter.cs:21]: Why check for null here? There is a null check in the Convert method

File: backend/api/Adapters/SubstructureAdapter.cs [hunk 2]
```diff
@@ -15,7 +15,16 @@ public static Substructure Convert(SubstructureDto substructureDto)
...
+
+            if (substructureDto.SubstructureCessasionCostProfileDto != null)
+            {
+                substructure.SubstructureCessasionCostProfile = Convert(substructureDto.SubstructureCessasionCostProfileDto, substructure);
+            }
```
Comment (taustad) [inline backend/api/Adapters/SubstructureAdapter.cs:26]: Why check for null here? There is a null check in the Convert method

File: backend/api/Adapters/SurfAdapter.cs [hunk 3]
```diff
@@ -20,7 +20,17 @@ public static Surf Convert(SurfDto surfDto)
                 ProductionFlowline = surfDto.ProductionFlowline,
                 RiserCount = surfDto.RiserCount
             };
-            surf.CostProfile = Convert(surfDto.CostProfile, surf);
+
+            if (surfDto.CostProfile != null)
+            {
+                surf.CostProfile = Convert(surfDto.CostProfile, surf);
+            }
```
Comment (taustad) [inline backend/api/Adapters/SurfAdapter.cs:27]: Why check for null here? There is a null check in the Convert method

File: frontend/src/Views/SubstructureView.tsx [hunk 4]
```diff
@@ -218,14 +294,48 @@ const SubstructureView = () => {
...
+            {(!costProfileDialogOpen
+                ? null : (
+                    <Import
+                        onClose={() => { setCostProfileDialogOpen(!costProfileDialogOpen) }}
+                        onImport={onCostProfileImport}
+                    />
+                ))}
+
+            <Wrapper>
+                <Typography variant="h4">Cessasion Cost profile</Typography>
```
Comment (taustad) [inline frontend/src/Views/SubstructureView.tsx:312]: Spelling

**Review by taustad — CHANGES_REQUESTED** [review]
Please fix "cessasion" -> "cessation"
Also remove redundant null checks in adapters
We should probably discuss using var instead of type for variables

File: backend/api/Adapters/SubstructureAdapter.cs (hunk 1 above)
Comment (ManveerPandher) [reply backend/api/Adapters/SubstructureAdapter.cs:21]: Fixed

File: backend/api/Adapters/SubstructureAdapter.cs (hunk 2 above)
Comment (ManveerPandher) [reply backend/api/Adapters/SubstructureAdapter.cs:26]: Fixed

File: backend/api/Adapters/SubstructureAdapter.cs [hunk 5]
```diff
@@ -47,5 +57,23 @@ public static void ConvertExisting(Substructure existing, SubstructureDto substr
...
             return substructureCostProfile;
         }
+
+        private static SubstructureCessasionCostProfile? Convert(SubstructureCessasionCostProfileDto? substructureCessasionCostProfileDto, Substructure substructure)
+        {
+            if (substructureCessasionCostProfileDto == null)
+            {
+                return null;
+            }
+            SubstructureCessasionCostProfile substructureCessasionCostProfile = new SubstructureCessasionCostProfile
```
[the remark this answers is not in the document]
Comment (ManveerPandher) [reply backend/api/Adapters/SubstructureAdapter.cs]: Another day

File: backend/api/Adapters/SurfAdapter.cs [hunk 6]
```diff
@@ -20,7 +20,17 @@ public static Surf Convert(SurfDto surfDto)
...
+
+            if (surfDto.SurfCessationCostProfileDto != null)
+            {
+                surf.SurfCessationCostProfile = Convert(surfDto.SurfCessationCostProfileDto, surf);
+            }
```
[the remark this answers is not in the document]
Comment (ManveerPandher) [reply backend/api/Adapters/SurfAdapter.cs:32]: Fixed

File: backend/api/Adapters/SurfAdapter.cs (hunk 3 above)
Comment (ManveerPandher) [reply backend/api/Adapters/SurfAdapter.cs:27]: Fixed

File: backend/api/Context/DcdDbContext.cs [hunk 7]
```diff
@@ -17,16 +17,20 @@ public DcdDbContext(DbContextOptions<DcdDbContext> options) : base(options)
...
+        public DbSet<SurfCessationCostProfile>? SurfCessationCostProfiles { get; set; }
 
         public DbSet<Substructure>? Substructures { get; set; }
         public DbSet<SubstructureCostProfile>? SubstructureCostProfiles { get; set; }
+        public DbSet<SubstructureCessasionCostProfile>? SubstructureCessasionCostProfiles { get; set; }
 
         public DbSet<Topside>? Topsides { get; set; }
 
         public DbSet<TopsideCostProfile>? TopsideCostProfiles { get; set; }
+        public DbSet<TopsideCessasionCostProfile>? TopsideCessasionCostProfiles { get; set; }
```
[the remark this answers is not in the document]
Comment (ManveerPandher) [reply backend/api/Context/DcdDbContext.cs]: Fixed

File: frontend/src/Views/SubstructureView.tsx [hunk 8]
```diff
@@ -96,20 +103,46 @@ const SubstructureView = () => {
                     setSubstructure(newSubstructure)
                 }
                 setSubstructureName(newSubstructure.name!)
-                const newColumnTitles = getColumnAbsoluteYears(caseResult, newSubstructure?.costProfile)
-                setColumns(newColumnTitles)
-                const newGridData = buildGridData(newSubstructure?.costProfile)
-                setGridData(newGridData)
+
+                const newCostProfileColumnTitles = getColumnAbsoluteYears(caseResult, newSubstructure?.costProfile)
+                const newCessasionCostProfileColumnTitles = getColumnAbsoluteYears(
```
[the remark this answers is not in the document]
Comment (ManveerPandher) [reply frontend/src/Views/SubstructureView.tsx]: Fixed

File: frontend/src/Views/SubstructureView.tsx (hunk 4 above)
Comment (ManveerPandher) [reply frontend/src/Views/SubstructureView.tsx]: Fixed

File: frontend/src/models/assets/substructure/Substructure.ts [hunk 9]
```diff
@@ -1,10 +1,12 @@
 import { SubstructureCostProfile } from "./SubstructureCostProfile"
+import { SubstructureCessasionCostProfile } from "./SubstructureCessasionCostProfile"
```
[the remark this answers is not in the document]
Comment (ManveerPandher) [reply frontend/src/models/assets/substructure/Substructure.ts]: Fixed

**Comment by ManveerPandher** [comment]
Fixed comments from QA

**Review by taustad — DISMISSED** [review]

**Review by petterwildhagen — APPROVED** [review]
LGTM! The few very minor issues can be addressed in separate PR (parenthesis, var instead of class name)

**Pull request merged**